src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java

```diff
@@ -0,0 +1,250 @@
+            return;
+        }
+        if(siteProvider.getSite(siteId) == null) {
+            ResponseUtil.error(rc, 404, "site_id: " + siteId + " not valid");
+            return;
+        }
+        String normalizedContactEmail = InputUtil.normalizeEmailString(contact);
+        if(normalizedContactEmail == null){
+            ResponseUtil.error(rc, 400, "contact email: " + contact + " not valid");
+            return;
+        }
+
+        final ClientSideKeypair newKeypair;
+        try {
+            newKeypair = createAndSaveSiteKeypair(siteId, contact, disabled);
+        } catch (Exception e) {
+            rc.fail(500, e);
+            return;
+        }
+        final JsonObject json = toJsonWithPrivateKey(newKeypair);
+        rc.response().putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                .end(json.encode());
+    }
+
+    private void handleUpdateKeypair(RoutingContext rc) {
+        final JsonObject body = rc.body().asJsonObject();
+        final String subscriptionId = body.getString("subscription_id");
+        final String contact = body.getString("contact");
+        Boolean disabled = body.getBoolean("disabled");
+
+        if(subscriptionId == null) {
+            ResponseUtil.error(rc, 400, "Required parameters: subscription_id");
+            return;
+        }
+
+        ClientSideKeypair keypair = this.keypairStore.getSnapshot().getKeypair(subscriptionId);
+        if(keypair == null) {
+            ResponseUtil.error(rc, 404, "Failed to find a keypair for subscription id: " + subscriptionId);
+            return;
+        }
+
+        if(contact == null && disabled == null) {
+            ResponseUtil.error(rc, 400, "Updatable parameters: contact, disabled");
+            return;
+        }
+        String normalizedContactEmail;
+        if (contact != null) {
+            normalizedContactEmail = InputUtil.normalizeEmailString(contact);
+            if(normalizedContactEmail == null){
+                ResponseUtil.error(rc, 400, "contact email: " + contact + " not valid");
+                return;
+            }
+        } else {
+            normalizedContactEmail = keypair.getContact();
+        }
+        if(disabled == null) {
+            disabled = keypair.isDisabled();
+        }
+
+        final ClientSideKeypair newKeypair = new ClientSideKeypair(
+                keypair.getSubscriptionId(),
+                keypair.encodePublicKeyToString(),
+                keypair.encodePrivateKeyToString(),
+                keypair.getSiteId(),
+                normalizedContactEmail,
+                keypair.getCreated(),
+                disabled);
+
+
+        Set<ClientSideKeypair> allKeypairs = new HashSet<>(this.keypairStore.getAll());
+        allKeypairs.remove(keypair);
+        allKeypairs.add(newKeypair);
+        try {
+            storeWriter.upload(allKeypairs, null);
+        }  catch (Exception e) {
+            rc.fail(500, e);
```

Comment:
- Does this log the exception?
- I wonder if all these try/catch/500s are necessary. The generic error handler may already be doing this.

Reply:
Looks like it can either be caught or thrown up to setupRoutes. If throw up to setupRoutes have to catch there since it's an overrride so can't change the signature to throw an exception. I think its best to keep it this way.

Reply:
Oh it's a checked exception? That's annoying. Fine for now, as long as these get logged

Reply:
logs explicit error message + the exception